manager/src/main/java/org/openremote/manager/mqtt/TeltonikaMQTTHandler.java

```diff
@@ -0,0 +1,639 @@
+            // TODO: Do we want to remove the value from TELTONIKA_DEVICE_SEND_COMMAND_ATTRIBUTE_NAME?
+            map.addAll(new Attribute<>
+                    (
+                            new AttributeDescriptor<>(DEVICE_RECEIVE_COMMAND_ATTRIBUTE_NAME, ValueType.TEXT),
+                            response.rsp
+                    ),
+                    new Attribute<>(
+                            new AttributeDescriptor<>(DEVICE_SEND_COMMAND_ATTRIBUTE_NAME, ValueType.TEXT),
+                    response.rsp
+                    )
+            );
+            return map;
+        }
+
+    }
+
+
+    /**
+     * Returns an {@code Optional<Attribute<AssetStateDuration>>}, that {@code ifPresent()}, represents
+     * the Duration for which the predicate returned true.
+     *
+     * @param previousValue The old attribute state (Or the latest datapoint that exists)
+     * @param newValue The new attribute value
+     * @param pred A Predicate that describes the state change
+     * @param ref An AttributeRef that describes which asset and attribute this pertains to.
+     * @return An Optional Attribute of type AssetStateDuration that represents the Duration for which the predicate returned true.
+     */
+    //TODO: Change this to only use an AttributeRef and a Predicate.
+    private Optional<Attribute<?>> assetChangedTripState(Attribute<?> previousValue, Attribute<?> newValue, AttributePredicate pred, AttributeRef ref) {
+        //We will first check if the predicate fails for the new value, and then check if the predicate is true for the previous value.
+        //In that way, we know that the state change happened between the new and previous values.
+
+        //TODO: Understand what happens with negate(), does it change states to the object itself?
+        boolean newValueTest =      pred.value.asPredicate(timerService::getCurrentTimeMillis).test(newValue        .getValue().get());
+        boolean previousValueTest = pred.value.asPredicate(timerService::getCurrentTimeMillis).test(previousValue   .getValue().get());
+        //If the predicate fails, then no changes need to happen.
+
+        // newValue is not 1, previousValue == 1
+        if(!(!newValueTest && previousValueTest)) {
+            return Optional.empty();
+        }
+
+        // Grab all datapoints (To be replaced by AssetDatapointValueQuery)
+        List<AssetDatapoint> valueDatapoints = AssetDatapointService.getDatapoints(ref);
+        ArrayList<AssetDatapoint> list = new ArrayList<>(valueDatapoints);
+
+        // If there are no historical data found, add some first
+        if(list.isEmpty()) return Optional.empty();
+
+        //What we do now is, we will try to figure out the latest datapoint where the predicate fails,
+        //before the newValue.
+        //This means that, the state change took place between the datapoint we just found and its next one.
+
+        //Find the first datapoint that passes the negated predicate
```

Comment:
Just as a practice; I would be hesitant in creating lots of multiline comments. Just for readability you know haha

Reply:
Maximum of 2 lines per comment, from now on, I hope nobody minds!